rust/template/differential_datalog/program.rs

```diff
@@ -1179,14 +1181,21 @@ impl Program {
                         let profcpu: &AtomicBool = &*profile_cpu3;
                         /* Filter out events we don't care about to avoid the overhead of sending
                          * the event around just to drop it eventually. */
-                        let mut filtered:Vec<((Duration, usize, TimelyEvent), String)> = data.drain(..).filter(|event| match event.2 {
+                        let mut filtered: Vec<(Duration, usize, TimelyEvent)> = data.drain(..).filter(|event| match event.2 {
                             TimelyEvent::Operates(_) => true,
-                            TimelyEvent::Schedule(_) => profcpu.load(Ordering::Acquire),
+                            TimelyEvent::Schedule(_) |
+                            TimelyEvent::GuardedMessage(_) |
+                            TimelyEvent::Park(_) |
+                            TimelyEvent::Progress(_) |
+                            TimelyEvent::PushProgress(_) => profcpu.load(Ordering::Acquire),
                             _ => false
-                        }).map(|x|(x, get_prof_context())).collect();
+                        }).collect();
+
                         if !filtered.is_empty() {
                             //eprintln!("timely event {:?}", filtered);
-                            prof_send1.send(ProfMsg::TimelyMessage(filtered.drain(..).collect())).unwrap();
+                            // Context doesn't change within a batch, so attach once.
```

Comment:
Before, the context String was attached to every message. From what I can tell, this string doesn't change at the "batch" granularity. Is this true? So I only include it once per batch (this allows the profile_statistics code to use the same vector as the existing profiling code `Vec<(Duration, usize, TimelyEvent)>`. Mainly:
```diff
- TimelyMessage(Vec<((Duration, usize, TimelyEvent), String)>)
+ TimelyMessage((Vec<(Duration, usize, TimelyEvent)>, String))